Upgrade tool (#1945)
strapi/documentation: docusaurus/docs/dev-docs/upgrade-tool.md

:::note
If your application is not already running the latest minor and patch version in the current major, the `major` upgrade is prevented, and you will first need to upgrade to the latest [minor.patch](#upgrade-to-a-minor-version) version in the current major version. This means that moving from v4.14.4 to v5.0.0 is a 2-step process because the latest v4 version is v4.16.2.
:::

## Upgrade to a minor version

Run the upgrade tool with the `minor` parameter to upgrade the project to the latest minor and patch version of Strapi:

```bash
npx @strapi/upgrade minor
```

During the upgrade process, the project dependencies are updated and installed, and the codemods up until the targeted patch are executed (if any).

## Upgrade to a patch version

Run the upgrade tool with the `patch` parameter to upgrade the project to the latest major version of Strapi:

```bash
npx @strapi/upgrade patch
```

During the upgrade process, the project dependencies are updated and installed, and the codemods up until the targeted patch are executed (if any).

> ## Run codemods only

Reviewer: This is still TBD for now.
This **should** be in the initial release but maybe we can add this later or add a "coming later/soon"?
What do you think?

Author: That's a good point. I would comment out these parts so they don't appear on the website, and whenever the feature is ready we would uncomment it. What do you think?

(If we're 100% sure it's going to be part of the feature for the stable v5 release, I could also add a callout mentioning something about codemods only is coming soon — I just don't want to disappoint users if in the end it's decided to be out of scope 😅)

Reviewer: Let's comment it out & add it back when the command will be there :slightly_smiling_face:

Run the upgrade tool with the `codemods` parameter to execute a utility that allows selecting the codemods to be executed. With this command, only the codemods are run, the dependencies are not updated nor installed.

```bash
npx @strapi/upgrade codemods
```

## Options

> The `npx @strapi/upgrade` command can accept the following options:

Reviewer: To nitpick a bit, the upgrade commands* can accept options, the upgrade CLI itself only has commands (major/minor/patch)

Author: Thank you, makes sense! I wasn't sure how to write but had a feeling it was not 100% correct :D

| Option                                                                   | Description                                                      |
| ------------------------------------------------------------------------ | ---------------------------------------------------------------- |

> | [`--dry-run`](#simulate-the-upgrade-without-updating-any-files-dry-run)  | Simulate the upgrade without updating any files (default: false) |

Reviewer: ```suggestion
| [`-n, --dry`](#simulate-the-upgrade-without-updating-any-files-dry-run)  | Simulate the upgrade without updating any files (default: false) |
```

| [`-d, --debug`](#get-detailed-debugging-information)                     | Get more logs in debug mode (default: false)                     |
| [`-s, --silent`](#execute-the-upgrade-silently)                          | Don't log anything (default: false)                              |

> | [`-p, --project-path`](#select-a-path-for-the-strapi-application-folder) | Path to the Strapi project                                       |

Reviewer: ```suggestion
| [`-p, --project-path <project-path>`](#select-a-path-for-the-strapi-application-folder) | Path to the Strapi project                                       |
```

| [`-V, --version`](#get-the-current-strapi-version)                       | Output the version number                                        |

> | [`-h, --help`](#get-help)                                                | Print command line options                                       |

Reviewer: Those commands can be run with or without commands before them, whereas the other ones need to be bound to commands. 

As they're more "utilities" do you think we should put them in another table/separate them from the rest? Or is it good like this?

Author: OK, thank you. I'll find a way to make it clear in the docs 👌

### Simulate the upgrade without updating any files (dry run)

> When passing the `--dry-run` option, the codemods are executed without actually editing the files. This allows simulating the upgrade of the codebase, checking the outcomes without applying any changes:

Reviewer: ```suggestion
When passing the `-n` or `--dry` option, the codemods are executed without actually editing the files. This allows simulating the upgrade of the codebase, checking the outcomes without applying any changes:
```

On top of that, the package.json won't be modified & the dependencies won't be re-installed

Examples:

```bash

> npx @strapi/upgrade major --dry-run
> npx @strapi/upgrade minor --dry-run
> npx @strapi/upgrade patch --dry-run
> npx @strapi/upgrade codemods --dry-run

Reviewer: ```suggestion
npx @strapi/upgrade major --dry
npx @strapi/upgrade minor --dry
npx @strapi/upgrade patch --dry
npx @strapi/upgrade codemods --dry
```

Also, same as above, do we want to document the `codemods` command now or should we wait/add a disclaimer?

```

### Select a path for the Strapi application folder

> When passing the `-p` option you can specify in which folder the Strapi application is located.

Reviewer: ```suggestion
When passing the `-p` or `--project-path` option followed by a valid path you can specify in which folder the Strapi application is located.
```

Example:

```bash
npx @strapi/upgrade major -p /path/to/the/Strapi/application/folder
```

### Get the current Strapi version

> When passing the `--version` option (or its `-V` shorthand), the current Strapi version is logged.

Reviewer: This will return the CLI version (the `@strapi/upgrade` package version), not Strapi's one

Example:

```sh
$ npx @strapi/upgrade -V
4.15.1
```

### Get detailed debugging information

When passing the `--debug` option (or its `-d` shorthand), the upgrade tool provides more detailed logs while running:

```bash
npx @strapi/upgrade --debug
```

### Execute the upgrade silently

When passing the `--silent` option (or its `-s` shorthand), the tool executes the upgrade without providing any log:

```bash
npx @strapi/upgrade --silent
```

### Get help

When passing the `--help` option (or its `-h` shorthand), help information is displayed, listing the available options:

Example:

```bash
$ npx @strapi/upgrade -h
Usage: upgrade [options]

Upgrade to the desired version

Options:
  --dry-run                          Simulate the upgrade without updating any files (default: false)
  -d, --debug                        Get more logs in debug mode (default: false)
  -s, --silent                       Don't log anything (default: false)
  -p, --project-path <project-path>  Path to the Strapi project
  -V, --version                      output the version number

>   -h, --help                         Print command line options

Reviewer: Using `upgrade --help` will display the CLI commands and options 

```text
Usage: upgrade <command> [options]

Options:
  -V, --version    output the version number
  -h, --help       Print command line options

Commands:
  major [options]  Upgrade to the next available major version of Strapi
  minor [options]  Upgrade to ...
  patch [options]  Upgrade to ...
  help [command]   Print options for a specific command
```

Whereas using `upgrade <command> --help` will display help for the actual command, displaying its options

```text
Usage: upgrade major [options]

Upgrade to the next available major version of Strapi

Options:
  -p, --project-path <project-path>  Path to the Strapi project
  -n, --dry                          Simulate the upgrade without updating any files (default: false)
  -d, --debug                        Get more logs in debug mode (default: false)
  -s, --silent                       Don't log anything (default: false)
  -h, --help                         display help for command
```

| [`-h, --help`](#get-help)                                                | Print command line options                                       |

### Simulate the upgrade without updating any files (dry run)

When passing the `-n` or `--dry` option, the codemods are executed without actually editing the files. The package.json will not be modified, and the dependencies will not be re-installed. Using this option allows simulating the upgrade of the codebase, checking the outcomes without applying any changes:

Examples:

```bash
npx @strapi/upgrade major --dry
npx @strapi/upgrade minor --dry
npx @strapi/upgrade patch --dry
```

### Select a path for the Strapi application folder

When passing the `-p` or `--project-path` option followed by a valid path you can specify in which folder the Strapi application is located.

Example:

```bash
npx @strapi/upgrade major -p /path/to/the/Strapi/application/folder
```

### Get the current version

When passing the `--version` option (or its `-V` shorthand), the current version of the upgrade tool is logged.

Example:

```sh
$ npx @strapi/upgrade -V

> 4.15.1

Author: I'm not sure the example output makes sense 🤔 I guess it would be more something like `1.0.1`? (i.e. the @strapi/upgrade version number is probably not aligned with the Strapi version number)

Reviewer: Actually it is

To clarify, it could indeed not be related at some point + it doesn't required it be in sync, etc, but it happens to be :man_shrugging:

```

### Get detailed debugging information

When passing the `--debug` option (or its `-d` shorthand), the upgrade tool provides more detailed logs while running:

```bash
npx @strapi/upgrade --debug
```

### Execute the upgrade silently

When passing the `--silent` option (or its `-s` shorthand), the tool executes the upgrade without providing any log:

```bash
npx @strapi/upgrade --silent
```

### Get help

When passing the `--help` option (or its `-h` shorthand), help information is displayed, listing the available options:

Examples:

<Tabs>
<Tab value="upgrade" label="General help for the upgrade tool">

```bash

> $ npx @strapi/upgrade -h

Author: Of course 🤦  Bad copy/paste… Thank you!

<ThemedImage
  alt="Version numbers explained"
  sources={{
    light: '/img/assets/update-migration/version-numbers.png',
    dark: '/img/assets/update-migration/version-numbers_DARK.png',
  }}
/>

- The first number is the **major** version number.
- The second number is the **minor** version number.
- The third number is the **patch** version number.

The upgrade tool allows upgrading to a major, minor, or patch version.

| The latest Strapi v4 version is v4.16.2 and my Strapi application is currently on… | If I run… | My Strapi application will be upgraded to … |
|----|----|----|
| v4.14.1 | `npx @strapi/upgrade patch` | v4.14.6<br/><br/>(because v4.14.6 is the latest patch version for the v4.14 minor version) |
| v4.14.1 | `npx @strapi/upgrade minor` | v4.16.2 |
| v4.14.1 | `npx @strapi/upgrade major` | Nothing.<br/><br/>I first need to run `npx @strapi/upgrade minor` to upgrade to v4.16.2. |
| v4.16.2 | `npx @strapi/upgrade major` | v5.0.0  |

## Upgrade to a major version

Run the upgrade tool with the `major` parameter to upgrade the project to the next major version of Strapi:

```bash
npx @strapi/upgrade major
```

> During the upgrade process, the application dependencies are updated and installed, and the codemods up until the targeted major version are executed.

Reviewer 2: since there's no "targeted major version" here, we should say "up until the next available major version" or something similar

(sorry, my Notion notes here were wrong and misleading)

:::note
If your application is not already running the latest minor and patch version in the current major, the `major` upgrade is prevented, and you will first need to upgrade to the latest [minor.patch](#upgrade-to-a-minor-version) version in the current major version. This means that moving from v4.14.4 to v5.0.0 is a 2-step process because the latest v4 version is v4.16.2.
:::

## Upgrade to a minor version

Run the upgrade tool with the `minor` parameter to upgrade the project to the latest minor and patch version of Strapi:

```bash
npx @strapi/upgrade minor
```

During the upgrade process, the project dependencies are updated and installed, and the codemods up until the targeted patch are executed (if any).

## Upgrade to a patch version

Run the upgrade tool with the `patch` parameter to upgrade the project to the latest major version of Strapi:

```bash
npx @strapi/upgrade patch
```

> During the upgrade process, the project dependencies are updated and installed, and the codemods up until the targeted patch are executed (if any).

Reviewer 2: and finally same here, "up until the latest patch for the current minor"
